### ISSUE TITLE
ContextNew callback failed. Failed to create new peer context

### ISSUE THREAD

**Issue by adityasood11** [FreeRDP/FreeRDP#4675]
>wfreerdp-server
[11:19:54:567] [25920:000073fc] [INFO][com.freerdp.server.windows] - screen id not provided. attempting to detect...
[11:19:54:567] [25920:000073fc] [INFO][com.freerdp.server.windows] - Detecting screens...
[11:19:54:569] [25920:000073fc] [INFO][com.freerdp.server.windows] - ID Resolution              Name (Interface)
[11:19:54:574] [25920:000073fc] [INFO][com.freerdp.server.windows] - 0  1536x864x32
[11:19:54:574] [25920:000073fc] [INFO][com.freerdp.server.windows] - \
[11:19:54:575] [25920:000073fc] [INFO][com.freerdp.server.windows] - Starting server
[11:19:54:578] [25920:000073fc] [INFO][com.freerdp.core.listener] - Listening on :::3389
[11:19:54:580] [25920:000073fc] [INFO][com.freerdp.core.listener] - Listening on 0.0.0.0:3389
[11:20:04:230] [25920:000037fc] [ERROR][com.freerdp.core.peer] - ContextNew callback failed
[11:20:04:234] [25920:000037fc] [ERROR][com.freerdp.core.peer] - Failed to create new peer context

This happens when i run wfreerdp-server and try to connect it with mstsc on 3389.

I generated the key using winpr-makecert.exe and its certificate using the command

`C:\OpenSSL-Win32\bin>openssl req -x509 -sha256 -nodes -days 365 -newkey rsa:2048 -keyout server.key -out server.crt`

and then i copied both key and crt to debug folder.

Note: I made some changes in code to remove the errors(undeclared identifiers) which i was getting while trying to build wfreerdp-server.

**Comment by adityasood11** [comment]
@akallabeth Atleast give a reason before closing.

**Comment by hardening** [comment]
@adityasood11 could it be related to the fact that you've not provided infos as we suggest in the issue template....

**Comment by adityasood11** [comment]
@hardening Can you please elaborate?

**Comment by adityasood11** [comment]
Steps to reproduce the error

Host System : Windows 10 64 bit 
Visual Studio 15 2017



Uncommented add_subdirectory("Windows") in CMakeLists.txt in freerdp-2.0.0-rc2/server
After building solution i got the following errors
1. TAG undeclared identifier ( which i rectified by adding TAG MACRO DEFINITION in wf_mirage.c)
2. Some variables were not members of SURFACE_BITS_COMMAND ( i rectified it using bmp)

Ex - cmd->codecID changed to cmd->bmp.codecID

3.  There were some more undeclared identifiers like WF_SRV_CALLBACK_EVENT  , WF_INFO_MAXPEERS etc

These identifiers had a different name in their definitions.

I changed WF_SRV_CALLBACK_EVENT_CONNECT  to FREERDP_SERVER_WIN_SRV_CALLBACK_EVENT_CONNECT

and 

WF_INFO_MAXPEERS to FREERDP_SERVER_WIN_INFO_MAXPEERS

etc.

**Comment by akallabeth** [comment]
@adityasood11 So you try to compile something that was intentionally deactivated and are surprised that that does not work?
You're using something that is mostly a stub/relic/unused/unmaintained so I'd suggest firing up your debugger and fix the issues you encounter until it works for you.
We're happy to integrate fixes for every part of the project.

Side note: we are by no means wizards that can answer every question without a problem description precise enough to at least guess what you're trying to achieve and some setup details (just windows vs2017  will not cut it)

**Comment by adityasood11** [comment]
I understand that. I wanted to share my issue so i can gain some insights on this from your team and i can work on my issue further.

**Comment by hardening** [comment]
@adityasood11 if you need help or want to ask questions please use IRC (#freerdp on freenode), this an issue tracker not a forum.

**Comment by adityasood11** [comment]
I have entered IRC through mIRC software using #freerdp.

**Comment by informatimago** [comment]
Note, the correct url for irc channels is like : irc://irc.freenode.org/#freerdp
This has nothing in common with the links for email: mailto:user@domain

**Comment by adityasood11** [comment]
Thanks @informatimago :)

**Comment by adityasood11** [comment]
I was looking in the above error. Using VS Debugger i was checking for possible reasons behind the error.
The error which is produced above is in wf_peer.c 

	```
IFCALLRET(client->ContextNew, ret, client, client->context);

	if (ret)
		return TRUE;

	WLog_ERR(TAG, "ContextNew callback failed");
fail_transport_attach:
	free(context->errorDescription);
fail_error_description:
	rdp_free(client->context->rdp);
fail_rdp:
	metrics_free(context->metrics);
fail_metrics:
	free(client->context);
fail_context:
	client->context = NULL;
	WLog_ERR(TAG, "Failed to create new peer context");
	return FALSE;
}
```

In the above code, ret variable should return 1 so that it will not enter the above section containing ContextNew Callback failed and Failed to create new peer context. But instead it is giving 0. 

I added a breakpoint on this 

`IFCALLRET(client->ContextNew, ret, client, client->context);`

When i check for the value of ret after running this project , i get the value as 1. 

Basically , when IFCALLRET is invoked, it takes value of ret as 1 but the value of ret which is returned is 0.


The definition of IFCALLRET is given in freerdp/api.h

`#define IFCALLRET(_cb, _ret, ...) do { if (_cb != NULL) { _ret = _cb( __VA_ARGS__ ); } } while (0)`

**Comment by dev0x2a** [comment]
Install DFMirage driver and run as admin.

**Comment by Harrywang55666** [comment]
@adityasood11 Have you solve the problem?
function `wf_mirror_driver_find_display_device ` called and it returns `FALSE`
seems string value `deviceInfo.DeviceString` returned from function`EnumDisplayDevices` does not equal `Mirage Driver`
```cpp
/*
This function will iterate over the loaded display devices until it finds
the mirror device we want to load. If found, it will then copy the registry
key corresponding to the device to the wfi and returns TRUE. Otherwise
the function returns FALSE.
*/
BOOL wf_mirror_driver_find_display_device(wfInfo* wfi)
{
	BOOL result;
	BOOL devFound;
	DWORD deviceNumber;
	DISPLAY_DEVICE deviceInfo;
	devFound = FALSE;
	deviceNumber = 0;
	deviceInfo.cb = sizeof(deviceInfo);

	while (result = EnumDisplayDevices(NULL, deviceNumber, &deviceInfo, 0))
	{
		if (_tcscmp(deviceInfo.DeviceString, _T("Mirage Driver")) == 0)
		{
			int deviceKeyLength;
			int deviceKeyPrefixLength;
			deviceKeyPrefixLength = _tcslen(DEVICE_KEY_PREFIX);

			if (_tcsnicmp(deviceInfo.DeviceKey, DEVICE_KEY_PREFIX, deviceKeyPrefixLength) == 0)
			{
				deviceKeyLength = _tcslen(deviceInfo.DeviceKey) - deviceKeyPrefixLength;
				wfi->deviceKey = (LPTSTR) malloc((deviceKeyLength + 1) * sizeof(TCHAR));

				if (!wfi->deviceKey)
					return FALSE;

				_tcsncpy_s(wfi->deviceKey, deviceKeyLength + 1,
				           &deviceInfo.DeviceKey[deviceKeyPrefixLength], deviceKeyLength);
			}

			_tcsncpy_s(wfi->deviceName, 32, deviceInfo.DeviceName, _tcslen(deviceInfo.DeviceName));
			return TRUE;
		}

		deviceNumber++;
	}

	return FALSE;
}
```
this is call stack
```
 wfreerdp-server.exe!wf_mirror_driver_find_display_device(wf_info * wfi) 48
 wfreerdp-server.exe!wf_mirror_driver_activate(wf_info * wfi) 328
 wfreerdp-server.exe!wf_info_peer_register(wf_info * wfi, wf_peer_context * context)  226
 wfreerdp-server.exe!wf_peer_context_new(rdp_freerdp_peer * client, wf_peer_context * context) 60	
 freerdp2.dll!freerdp_peer_context_new(rdp_freerdp_peer * client) 795
 wfreerdp-server.exe!wf_peer_init(rdp_freerdp_peer * client)  90
 wfreerdp-server.exe!wf_peer_main_loop(void * lpParam)  250
```

my Windows7 is inside of VMWare.
The items of the list are what `deviceInfo.DeviceString` from `EnumDisplayDevices`
* VMware SVGA 3D
* RDPDD Chained DD
* RDP Encoder Mirror Driver
* RDP Reflector Display Driver

@dev0x2a Is there any another way to solve this problem besides installing DFMirage driver

**Comment by Zoson** [comment]
@adityasood11 Had you solved the problem? This error also bother me !

**Comment by Zoson** [comment]
@Harrywang55666 Hi， Did you resolve the error?

**Comment by Zoson** [comment]
@Harrywang55666 ，The error occured when the code ran into <wf_mirror_driver_map_memory> function in wf_mirage.c, and it couldn't createDC with deviceNname "\\.\DISPLAYV1".
